versioned_hdf5/wrappers.py

```diff
@@ -200,63 +204,92 @@ def create_group(self, name, track_order=None):
+            raise NotImplementedError("Empty datasets are not supported.")
+
+        if data is not None:
+            if not isinstance(data, np.ndarray) or data.dtype.kind == "O":
+                dtype = dtype or guess_dtype(data)
+            data = np.asarray(data, order="C", dtype=dtype)
+
         if shape is None:
+            if data is None:
+                raise TypeError("Either shape or data must be specified")
             shape = data.shape
-        if fillvalue is not None and isinstance(data, np.ndarray):
-            data = InMemoryArrayDataset(name, data, parent=self, fillvalue=fillvalue)
-        chunks = kwds.get("chunks")
-        if data is None:
-            data = InMemorySparseDataset(
-                name,
-                shape=shape,
-                dtype=dtype,
-                parent=self,
-                fillvalue=fillvalue,
-                chunks=chunks,
+        elif data is not None and data.shape != shape:
+            raise ValueError(
+                f"data.shape {data.shape} does not match specified shape {shape}"
             )
-        if chunks in [True, None]:
+
+        if chunks in (True, None):
             if len(shape) == 1:
                 chunks = (DEFAULT_CHUNK_SIZE,)
             else:
                 raise NotImplementedError(
                     "chunks must be specified for multi-dimensional datasets"
                 )
-        if isinstance(chunks, int) and not isinstance(chunks, bool):
+        elif chunks is False:
+            raise ValueError("chunks=False is not supported for versioned datasets")
+        elif isinstance(chunks, int) and not isinstance(chunks, bool):
             chunks = (chunks,)
+        assert isinstance(chunks, tuple)
+
         if len(shape) != len(chunks):
             raise ValueError(
-                f"Dimensions of chunks ({chunks}) must equal the dimensions of the shape ({shape})"
+                f"Dimensions of chunks ({chunks}) must equal the dimensions of the "
+                f"shape ({shape})"
             )
         if len(shape) == 0:
             raise NotImplementedError("Scalar datasets are not implemented.")
 
+        ds: DatasetLike
+        if data is not None:
+            ds = InMemoryArrayDataset(
+                name,
+                data,
+                parent=self,
+                fillvalue=fillvalue,
+                chunks=chunks,
+            )
+        else:
+            ds = InMemorySparseDataset(
+                name,
+                shape=shape,
+                dtype=dtype,
+                parent=self,
+                fillvalue=fillvalue,
+                chunks=chunks,
+            )
+
         self.set_chunks(name, chunks)
-        self.set_compression(name, kwds.get("compression"))
-        self.set_compression_opts(name, kwds.get("compression_opts"))
-        self[name] = data
-        if dtype is not None:
-            self[name]._dtype = dtype
-        return self[name]
```

Comment:
`data` could be a np.ndarray in some use cases at this point. Took me a while to figure them out.